### PR TITLE
[ISSUE #714] Add alterTable testUtil And Fix partition error message

### PR DESCRIPTION
**Pull request by shawnding**
[ISSUE #714] Add alterTable testUtil And Fix partition error message

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/TableMetadata.java [hunk 1]
```diff
@@ -589,4 +590,20 @@ private static PartitionSpec freshSpec(int specId, Schema schema, PartitionSpec
...
+  public void checkCompatibility(Schema newSchema) {
+    for (PartitionField field : spec().fields()) {
+      Type sourceType = newSchema.findType(field.sourceId());
+      if (sourceType == null) {
+        throw new UnsupportedOperationException("Cannot support drop the partition column: " + field.name());
+      }
+      if (!sourceType.isPrimitiveType()) {
+        throw new UnsupportedOperationException("Cannot partition by non-primitive source field: " + field.name());
+      }
+      if (!field.transform().canTransform(sourceType)) {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/TableMetadata.java]: Please add newlines between these if statements.

**Comment by rdblue** [comment]
I think it would be better to catch this while updating the schema, rather than after running the schema update and trying to commit it. That way, `UpdateSchema.apply` will also catch the failure.

Can you add precondition checks to `UpdateSchema.deleteColumn`? And tests should be added to `TestSchemaUpdate`.

**Comment by shawnding** [comment]
HI Ryan, I had add precondition checks to UpdateSchema.deleteColumn.  Because  the table spec should base on the table not only schema, so I Add the test util in HiveTableTest.java

File: core/src/main/java/org/apache/iceberg/SchemaUpdate.java [hunk 2]
```diff
@@ -150,6 +150,10 @@ public UpdateSchema deleteColumn(String name) {
     Preconditions.checkArgument(!updates.containsKey(field.fieldId()),
         "Cannot delete a column that has updates: %s", name);
 
+    if (base != null && base.spec().getFieldsBySourceId(field.fieldId()).size() != 0) {
+      throw new UnsupportedOperationException("Cannot support to drop the partition column: " + name);
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/SchemaUpdate.java]: This should be a precondition check like the other validations.

Let's change the error message to `"Cannot drop column used by partition: %s(%s)", partitionField.transform(), name`

**Comment by rdblue** [comment]
@shawnding, I'm not sure I understand your rationale for adding the test to HiveTableTest.

At a minimum, this should be a new test case and should not alter an unrelated test. It's okay to have a new test case in HiveTableTest, but i don't think this needs to modify the test base. Also, this needs to be tested in `TestSchemaUpdate`.

**Comment by rdblue** [comment]
@shawnding, could you update the tests? I think we can get this one in soon.

**Comment by rdblue** [comment]
@shawnding, do you intend to continue working on this?

**Comment by rdblue** [comment]
Since there hasn't been an update since January, I'm going to close this. Feel free to reopen if you want to pick it up again.

**Pull request closed without merging**